[UX-33] rpk: migrate login + profile creation flow to the Public API.

r-vasquez opened the pull request:
This PR migrates the login + profile creation flow to the Public API client instead of the Cloud API. The end goal is to migrate away from the CloudAPI client.

There is no change for the end user and the commands remain the same. These are just internal changes.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes
* none

r-vasquez committed:
rpk: change Organization fetch to publicAPI

This is done in the oauth.loadFlow of the normal
login steps.

Also renamed the publicapi.ControlPlaneClientSet
to CloudClientSet since the clients can come from
different services and not only the "controlplane"
service.

r-vasquez committed:
rpk: migrate login/profile creation flow to PublicAPI

This is an ongoing migration to the new Cloud API.
There is no difference for the end user.

andresaristizabal commented:
checking PR

andresaristizabal commented on src/go/rpk/pkg/publicapi/controlplane.go:
```diff
@@ -80,3 +145,88 @@ func (cpCl *ControlPlaneClientSet) ClusterForID(ctx context.Context, ID string)
+// which have pagination enabled.
+func Paginate[T any](
+	ctx context.Context,
+	maxPages int,
```
Is max pages necessary? why is nextTokenPage == "" not enough?

r-vasquez replied:
Just being paranoid to avoid an infinite loop. 

Basically to avoid the case of a rogue server that never stops sending the `nextTokenPage` with the last saved value.

andresaristizabal commented on src/go/rpk/pkg/publicapi/controlplane.go:
```diff
@@ -66,9 +73,67 @@ func (cpCl *ControlPlaneClientSet) ResourceGroupForID(ctx context.Context, ID st
+func (cpCl *CloudClientSet) ResourceGroups(ctx context.Context) ([]*controlplanev1beta2.ResourceGroup, error) {
+	maxPages := 200
+	fetchPage := func(ctx context.Context, pageToken string) ([]*controlplanev1beta2.ResourceGroup, string, error) {
+		req := connect.NewRequest(&controlplanev1beta2.ListResourceGroupsRequest{PageToken: pageToken})
```
what do you think add `PageSize: 100`, in order to avoid default values which is 50.

r-vasquez replied:
Oh, that's great, I'll add that.

andresaristizabal commented on src/go/rpk/pkg/publicapi/controlplane.go:
```diff
@@ -66,9 +73,67 @@ func (cpCl *ControlPlaneClientSet) ResourceGroupForID(ctx context.Context, ID st
+	return Paginate(ctx, maxPages, fetchPage)
+}
+
+func (cpCl *CloudClientSet) ServerlessClusterForID(ctx context.Context, ID string) (*controlplanev1beta2.ServerlessCluster, error) {
```
[nit]: I don't know if we have a standard in rpk, but, what do you think call it `GetServerlessClusterByID`?

r-vasquez replied:
Not an rpk standard but instead of Go, we try to be as close as possible to general naming conventions.

For this case in particular: [Getters](https://go.dev/doc/effective_go#Getters).

**rpk instances where we follow this (some examples):**
- https://github.com/redpanda-data/redpanda/blob/dev/src/go/rpk/pkg/cloudapi/api_installpack.go#L54
- https://github.com/redpanda-data/redpanda/blob/dev/src/go/rpk/pkg/cloudapi/api_organization.go#L26

**rpk instances where we don't follow this:**
to fulfill interfaces:
- https://github.com/redpanda-data/redpanda/blob/dev/src/go/rpk/pkg/publicapi/transform.go#L62
to add extra logic/defaults :
- https://github.com/redpanda-data/redpanda/blob/dev/src/go/rpk/pkg/config/rpk_yaml.go (However, one could argue this could be named differently)

However, there are some other places where we have failed to follow this too in the past.

andresaristizabal commented on src/go/rpk/pkg/publicapi/controlplane.go:
```diff
@@ -66,9 +73,67 @@ func (cpCl *ControlPlaneClientSet) ResourceGroupForID(ctx context.Context, ID st
+
+// Clusters returns all the Clusters using the pagination feature to traverse
+// all pages of the list.
+func (cpCl *CloudClientSet) Clusters(ctx context.Context) ([]*controlplanev1beta2.Cluster, error) {
```
[nit]: same here "ListClusters" or "GetClusters"

r-vasquez replied:
Same as above.

andresaristizabal commented on src/go/rpk/pkg/publicapi/controlplane.go:
```diff
@@ -66,9 +73,67 @@ func (cpCl *ControlPlaneClientSet) ResourceGroupForID(ctx context.Context, ID st
+func (cpCl *CloudClientSet) Clusters(ctx context.Context) ([]*controlplanev1beta2.Cluster, error) {
+	maxPages := 500
+	fetchPage := func(ctx context.Context, pageToken string) ([]*controlplanev1beta2.Cluster, string, error) {
+		req := connect.NewRequest(&controlplanev1beta2.ListClustersRequest{PageToken: pageToken})
```
I think you can add PageSize here.

andresaristizabal commented on src/go/rpk/pkg/publicapi/controlplane.go:
```diff
@@ -80,3 +145,88 @@ func (cpCl *ControlPlaneClientSet) ClusterForID(ctx context.Context, ID string)
+	wg.Wait()
+
+	err := errors.Join(orgErr, rgsErr, vcErr, cErr)
+	return org, rgs, vcs, cs, err
```
is it safe to use var out of the go routine?

r-vasquez replied:
In **this** case, it is. We are not modifying any variable concurrently, every pair of variables is modified separately in different goroutines, and we don't need any synchronization (mutex for example).

For context, this is the equivalent function of our current Cloud API helper function in: https://github.com/redpanda-data/redpanda/blob/dev/src/go/rpk/pkg/cloudapi/cloudapi.go#L66 but using the Public API

andresaristizabal commented on src/go/rpk/pkg/publicapi/controlplane.go:
```diff
@@ -41,17 +46,19 @@ func NewControlPlaneClientSet(host, authToken string, opts ...connect.ClientOpti
+	return &CloudClientSet{
 		Cluster:       controlplanev1beta2connect.NewClusterServiceClient(httpCl, host, opts...),
-	}, nil
+		Organization:  iamv1beta2connect.NewOrganizationServiceClient(httpCl, host, opts...),
```
How about passing those clients as parameters to NewCloudClientSet? This approach allows us to use mockgen to generate mocks for these clients, making tests easier to write and clearer to understand. Additionally, we can leverage protobuf to auto-generate the necessary mocks.

r-vasquez replied:
Yes, having a way to inject the internal clients can be beneficial as we expand the PublicAPI usage, especially when we introduce more convoluted usage such as in creating a Cluster.

What we can do is to refactor the client creation to accept optional client configuration options, and default to our current implementation, this is what I'm thinking:

```
type CloudClientSet struct {
	Cluster       controlplanev1beta2connect.ClusterServiceClient
	Organization  iamv1beta2connect.OrganizationServiceClient
	ResourceGroup controlplanev1beta2connect.ResourceGroupServiceClient
	Serverless    controlplanev1beta2connect.ServerlessClusterServiceClient
}

type CloudClientSetOptions struct {
	connectOptions []connect.ClientOption
	httpClient     *http.Client

	Cluster       controlplanev1beta2connect.ClusterServiceClient
	Organization  iamv1beta2connect.OrganizationServiceClient
	ResourceGroup controlplanev1beta2connect.ResourceGroupServiceClient
	Serverless    controlplanev1beta2connect.ServerlessClusterServiceClient
}

// Opt configures the CloudClientSet options.
type Opt func(*CloudClientSetOptions)

func NewCloudClientSet(host, authToken string, opts ...Opt) *CloudClientSet {
	if host == "" {
		host = ControlPlaneProdURL
	}

        // Default values
	clOpts := &CloudClientSetOptions{
		connectOptions: []connect.ClientOption{
			connect.WithInterceptors(
				newAuthInterceptor(authToken), // Add the Bearer token.
				newLoggerInterceptor(),        // Add logs to every request.
			),
		},
		httpClient: &http.Client{Timeout: 30 * time.Second},
	}
	// Apply custom options
	for _, opt := range opts {
		opt(clOpts)
	}

	// Initialize the CloudClientSet with default clients
	cl := &CloudClientSet{
		Cluster:       controlplanev1beta2connect.NewClusterServiceClient(clOpts.httpClient, host, clOpts.connectOptions...),
		Organization:  iamv1beta2connect.NewOrganizationServiceClient(clOpts.httpClient, host, clOpts.connectOptions...),
		ResourceGroup: controlplanev1beta2connect.NewResourceGroupServiceClient(clOpts.httpClient, host, clOpts.connectOptions...),
		Serverless:    controlplanev1beta2connect.NewServerlessClusterServiceClient(clOpts.httpClient, host, clOpts.connectOptions...),
	}

	// Override with custom clients from options if provided
	if clOpts.Cluster != nil {
		cl.Cluster = clOpts.Cluster
	}
	if clOpts.Organization != nil {
		cl.Organization = clOpts.Organization
	}
	if clOpts.ResourceGroup != nil {
		cl.ResourceGroup = clOpts.ResourceGroup
	}
	if clOpts.Serverless != nil {
		cl.Serverless = clOpts.Serverless
	}
	return cl
}

// Options setters:

func WithClusterServiceClient(cluster controlplanev1beta2connect.ClusterServiceClient) Opt {
	return func(cl *CloudClientSetOptions) { cl.Cluster = cluster }
}

func WithOrganizationClient(organization iamv1beta2connect.OrganizationServiceClient) Opt {
	return func(cl *CloudClientSetOptions) { cl.Organization = organization }
}

func WithResourceGroupClient(resourceGroup controlplanev1beta2connect.ResourceGroupServiceClient) Opt {
	return func(cl *CloudClientSetOptions) { cl.ResourceGroup = resourceGroup }
}

func WithServerlessClient(serverless controlplanev1beta2connect.ServerlessClusterServiceClient) Opt {
	return func(cl *CloudClientSetOptions) { cl.Serverless = serverless }
}

func WithConnectClientOptions(opts ...connect.ClientOption) Opt {
	return func(cl *CloudClientSetOptions) { cl.connectOptions = append(cl.connectOptions, opts...) }
}

func WithHTTPClient(client *http.Client) Opt {
	return func(cl *CloudClientSetOptions) { cl.httpClient = client }
}
```

However, such a refactor does not currently benefit us, as we don't currently need to mock any of our tests. So, I'll leave this as a reference for any future refactor for this client.

I've added a fake response to one of our tests as you suggested earlier: https://github.com/redpanda-data/redpanda/pull/24877/commits/b142aa2f441db583c1697eac514720c499e59aca

r-vasquez committed:
rpk: add response test server for oauth.Load

Instead of setting the inTest boolean on the test
and having testing code in the implementation.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--fd15394e-20c2-4471-8872-0197aa8abeb5-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61325">build#61325</a></summary>


| test_id                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   |
|:---------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61325#0194b0a8-f6ca-4cba-ae0c-5c725fe9a641 | FLAKY         | 1/2      |
| rptest.tests.timequery_test.TimeQueryKafkaTest.test_timequery_below_start_offset | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61325#0194b0a8-f6cc-48ca-b061-3b4b3a9acef4 | FLAKY         | 1/2      |
</details>
<!--/body-->

andresaristizabal commented on src/go/rpk/pkg/oauth/load_test.go:
```diff
@@ -93,6 +94,25 @@ func TestLoadFlow(t *testing.T) {
 			}
+			handler := func() http.HandlerFunc {
+				return func(w http.ResponseWriter, r *http.Request) {
+					if strings.Contains(r.URL.Path, "GetCurrentOrganization") {
```
NICE!

r-vasquez commented:
/backport v24.3.x

r-vasquez commented:
/backport v24.2.x

vbotbuildovich commented:
Failed to create a backport PR to v24.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-24877-v24.2.x-231 remotes/upstream/v24.2.x
git cherry-pick -x 219adb2dbf ecfbbc8d1f b142aa2f44
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/13552769348).
